libraries/hotstuff/qc_chain.cpp

```diff
@@ -566,32 +569,30 @@ namespace eosio::hotstuff {
          _pending_proposal_block = {};
          _b_leaf = proposal_candidate.proposal_id;
 
-         send_hs_proposal_msg(proposal_candidate);
+         send_hs_msg(proposal_candidate);
 
          fc_tlog(_logger, " === ${id} _b_leaf updated (on_beat): ${proposal_id}", ("proposal_id", proposal_candidate.proposal_id)("id", _id));
       }
    }
 
-   void qc_chain::send_hs_proposal_msg(const hs_proposal_message & msg){
-      fc_tlog(_logger, " === broadcast_hs_proposal ===");
-      _pacemaker->send_hs_proposal_msg(msg, _id);
-      process_proposal(msg);
-   }
-
-   void qc_chain::send_hs_vote_msg(const hs_vote_message & msg){
-      fc_tlog(_logger, " === broadcast_hs_vote ===");
-      _pacemaker->send_hs_vote_msg(msg, _id);
-      process_vote(msg);
-   }
-
-   void qc_chain::send_hs_new_view_msg(const hs_new_view_message & msg){
-      fc_tlog(_logger, " === broadcast_hs_new_view ===");
-      _pacemaker->send_hs_new_view_msg(msg, _id);
-   }
-
-   void qc_chain::send_hs_new_block_msg(const hs_new_block_message & msg){
-      fc_tlog(_logger, " === broadcast_hs_new_block ===");
-      _pacemaker->send_hs_new_block_msg(msg, _id);
+   std::optional<hs_commitment> qc_chain::send_hs_msg(const hs_message& msg) {
```

Comment:
Where is it copied?

Reply:
But yes we create a larger variant for some messages.

Reply:
When you create a variant from a type it is copied unless you move into the construction. Might want to take by value and move into it to avoid the copy. Although, most of the data in the messages is likely not movable.